cpp/include/cudf/detail/algorithm/reduce.cuh

```diff
@@ -136,5 +140,88 @@ OutputType transform_reduce(InputIterator begin,
+ * @tparam ValuesOutputIterator **[inferred]** Type of device-accessible output iterator for reduced
+ * values
+ *
+ * @param keys_begin Device-accessible iterator to start of input keys
+ * @param keys_end Device-accessible iterator to end of input keys
+ * @param values_begin Device-accessible iterator to start of input values
+ * @param keys_output Device-accessible iterator to start of output unique keys
+ * @param values_output Device-accessible iterator to start of output reduced values
+ * @param op Binary reduction operator
+ * @param stream CUDA stream to use
+ * @return A pair of iterators pointing to the end of the output key and value ranges
+ */
+template <typename Op,
+          typename KeysInputIterator,
+          typename KeysOutputIterator,
+          typename ValuesInputIterator,
+          typename ValuesOutputIterator>
+cuda::std::pair<KeysOutputIterator, ValuesOutputIterator> reduce_by_key(
+  KeysInputIterator keys_begin,
+  KeysInputIterator keys_end,
+  ValuesInputIterator values_begin,
+  KeysOutputIterator keys_output,
+  ValuesOutputIterator values_output,
+  Op op,
+  rmm::cuda_stream_view stream)
+{
+  auto const num_items = cuda::std::distance(keys_begin, keys_end);
+
+  // Device memory to store the number of runs (unique keys)
+  auto d_num_runs =
+    rmm::device_uvector<cuda::std::size_t>(1, stream, cudf::get_current_device_resource_ref());
+
+  // First call to get temporary storage size
+  size_t temp_storage_bytes = 0;
+  cub::DeviceReduce::ReduceByKey(nullptr,
+                                 temp_storage_bytes,
+                                 keys_begin,
+                                 keys_output,
+                                 values_begin,
+                                 values_output,
+                                 d_num_runs.begin(),
+                                 op,
+                                 num_items,
+                                 stream.value());
+
+  // Allocate temporary storage
+  rmm::device_buffer d_temp_storage(
+    temp_storage_bytes, stream, cudf::get_current_device_resource_ref());
+
+  // Run reduce-by-key
+  cub::DeviceReduce::ReduceByKey(d_temp_storage.data(),
+                                 temp_storage_bytes,
+                                 keys_begin,
+                                 keys_output,
+                                 values_begin,
+                                 values_output,
+                                 d_num_runs.begin(),
+                                 op,
+                                 num_items,
+                                 stream.value());
+
+  // Copy number of runs back to host via pinned memory
+  auto num_runs = cudf::detail::make_pinned_vector<size_t>(size_t{1}, stream);
+  cudf::detail::cuda_memcpy(
+    cudf::host_span<cuda::std::size_t>{&num_runs.front(), size_t{1}},
+    cudf::device_span<cuda::std::size_t const>{d_num_runs.begin(), size_t{1}},
+    stream);
```

Comment:
Can we implement a factory function to perform the memcpy for the user?  Similar to https://github.com/rapidsai/cudf/blob/d3fc48a1a0b9291ae6bac21af27e1039e63ba3cf/cpp/include/cudf/detail/utilities/vector_factories.hpp#L377-L383

And: https://github.com/rapidsai/cudf/blob/d3fc48a1a0b9291ae6bac21af27e1039e63ba3cf/cpp/include/cudf/detail/utilities/vector_factories.hpp#L418-L424